compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -6760,6 +6760,68 @@ TR::Node *laddSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    return node;
    }
 
+TR::Node *fpAddSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
+   {
+   TR::Node *firstChild, *secondChild;
+   if (TR::Compiler->arith.permitsFusedMultiplyAdd())
+      {
+      firstChild = node->getFirstChild();
+      secondChild = node->getSecondChild();
+      if (s->cg()->supportsFusedMultiplyAdd() && firstChild->getOpCode().isMul())
+         {
+         if (performTransformation(s->comp(), "%sReduced [f/d]add with first child [f/d]mul in node [%s] to [f/d]muladd\n", s->optDetailString(), node->getName(s->getDebug())))
+            {
+            TR::Node *newNode;
+            if (node->getOpCode().isFloat())
+               newNode = TR::Node::create(node, TR::fmuladd, 3);
+            else
+               newNode = TR::Node::create(node, TR::dmuladd, 3);
+            newNode->setAndIncChild(2, secondChild);
+            newNode->setAndIncChild(1, firstChild->getSecondChild());
+            newNode->setAndIncChild(0, firstChild->getFirstChild());
+            node->recursivelyDecReferenceCount();
+            newNode->incReferenceCount();
+            s->_alteredBlock = true;
+            return newNode;
+            }
+         }
+      else if (firstChild->getOpCode().isMulAdd() &&
+               firstChild->getChild(2)->getOpCode().isLoadConst() && 
+               ((firstChild->getOpCode().isFloat() && firstChild->getChild(2)->getFloat() == 0.0) ||
+                (firstChild->getOpCode().isDouble() && firstChild->getChild(2)->getDouble() == 0.0)))
+         {
+         if (performTransformation(s->comp(), "%sReduced [f/d]add with first child fma(a, b, 0) in node [%s] to fma(a, b, c)\n", s->optDetailString(), node->getName(s->getDebug())))
+            {
+            TR::Node *newNode;
+            if (firstChild->getOpCodeValue() == TR::fmuladd || firstChild->getOpCodeValue() == TR::fmulsub)
```

Comment:
I think this would look cleaner as a switch and I'd like to see a default case asserting.